gax/src/main/java/com/google/api/gax/batching/BatcherStats.java

```diff
@@ -0,0 +1,194 @@
+
+    synchronized (lock) {
+      if (throwable instanceof ApiException) {
+        StatusCode.Code code = ((ApiException) throwable).getStatusCode().getCode();
+        exceptionClass = ApiException.class;
+
+        if (requestStatusCounts.containsKey(code)) {
+          Integer codeCount = requestStatusCounts.get(code);
+          requestStatusCounts.put(code, ++codeCount);
+
+        } else {
+          requestStatusCounts.put(code, 1);
+        }
+      }
+
+      if (requestExceptionCounts.containsKey(exceptionClass)) {
+        Integer exCount = requestExceptionCounts.get(exceptionClass);
+        requestExceptionCounts.put(exceptionClass, ++exCount);
+      } else {
+        requestExceptionCounts.put(exceptionClass, 1);
+      }
+    }
+  }
+
+  <T> void recordBatchElementsCompletion(List<SettableApiFuture<T>> batchElementResultFutures) {
+    final AtomicBoolean markBatchFailure = new AtomicBoolean();
+
+    for (ApiFuture<T> elementResult : batchElementResultFutures) {
+      ApiFutures.addCallback(
+          elementResult,
+          new ApiFutureCallback<T>() {
+            @Override
+            public void onFailure(Throwable throwable) {
+              if (markBatchFailure.compareAndSet(false, true)) {
+                partialBatchFailures.incrementAndGet();
+              }
+              Class exceptionClass = throwable.getClass();
+
+              synchronized (lock) {
+                if (throwable instanceof ApiException) {
+                  StatusCode.Code code = ((ApiException) throwable).getStatusCode().getCode();
+                  exceptionClass = ApiException.class;
+
+                  if (entryStatusCounts.containsKey(code)) {
+                    Integer statusCount = entryStatusCounts.get(code);
+                    entryStatusCounts.put(code, ++statusCount);
+                  } else {
+                    entryStatusCounts.put(code, 1);
+                  }
+                }
+
+                if (entryExceptionCounts.containsKey(exceptionClass)) {
+                  Integer exCount = entryExceptionCounts.get(exceptionClass);
+                  entryExceptionCounts.put(exceptionClass, ++exCount);
+                } else {
+                  entryExceptionCounts.put(exceptionClass, 1);
+                }
+              }
+            }
+
+            @Override
+            public void onSuccess(T result) {}
+          },
+          directExecutor());
+    }
+  }
+
+  /** Calculates and formats the message with request and entry failure count. */
+  @Nullable
+  BatchingException asException() {
+    synchronized (lock) {
+      int partialFailures = partialBatchFailures.get();
+      if (requestExceptionCounts.isEmpty() && partialFailures == 0) {
+        return null;
+      }
+
+      StringBuilder sb = new StringBuilder();
+      int batchFailures = requestExceptionCounts.size();
+
+      if (requestExceptionCounts.isEmpty()) {
+        sb.append("Batching finished with ");
```

Comment:
I think the 2 branch should have the same message prefix:

"Batching finished with %d failed batches due to: .... %d NullPointerExceptions, %d ApiExceptions (%d DEADLINE_EXCEEDED) and %d partial failures. Partial failures ..."
"Batching finished with %d partial failures: ... . Parital failures ..."